File: problems/800-similar-rgb-color/enum.js
/**
 * 枚举：枚举全部可简化的颜色
 */
var similarRGB = function(color) {
  const target = color2RGB(color);
  const step = 16 * 1 + 1;
  let res = '';
  let minDiff = Infinity; // 当前结果`res`与`color`的距离

  for (let r = 0; r <= 255; r += step) {
    for (let g = 0; g <= 255; g += step) {
      for (let b = 0; b <= 255; b += step) {
        const d = Math.pow(r - target[0], 2) + Math.pow(g - target[1], 2) + Math.pow(b - target[2], 2);
        if (d < minDiff) { // 如果差距更小，则更新答案
          minDiff = d;
          res = rgb2Color(r, g, b);
        }
      }
    }
  }

  return res;
};

function color2RGB(color) {
  const res = [];
  res.push(Number.parseInt(color.slice(1, 3), 16));
  res.push(Number.parseInt(color.slice(3, 5), 16));
  res.push(Number.parseInt(color.slice(5, 7), 16));
  return res;
}

function rgb2Color(r, g, b) {
  let res = '#';
  res += r ? r.toString(16) : '00';
  res += g ? g.toString(16) : '00';
  res += b ? b.toString(16) : '00';
  return res;
}

File: problems/800-similar-rgb-color/enum.cpp
/**
 * 枚举：枚举所有可简写的颜色
 */

class Solution {
public:
    string similarRGB(string color) {
        auto targetRgb = color2Rgb(color);
        int minDiff = INT_MAX;
        string minColor;

        for (int r = 0; r < 16; ++r) {
            for (int g = 0; g < 16; ++g) {
                for (int b = 0; b < 16; ++b) {
                    int red = r * 16 + r;
                    int green = g * 16 + g;
                    int blue = b * 16 + b;

                    int diff = (red - targetRgb[0]) * (red - targetRgb[0]) +
                        (green - targetRgb[1]) * (green - targetRgb[1]) +
                        (blue - targetRgb[2]) * (blue - targetRgb[2]);
                    if (diff < minDiff) {
                        minDiff = diff;
                        minColor = rgb2Color({ red, green, blue });
                    }
                }
            }
        }

        return minColor;
    }

    string rgb2Color(vector<int> rgb) {
        string res("#");
        for (int n : rgb) {
            res += decimal2Hex(n);
        }
        return res;
    }

    vector<int> color2Rgb(string &color) {
        vector<int> res;
        res.push_back(hex2Decimal(color.substr(1, 2)));
        res.push_back(hex2Decimal(color.substr(3, 2)));
        res.push_back(hex2Decimal(color.substr(5, 2)));
        return res;
    }

    string decimal2Hex(int n) {
        if (n == 0) return "00";

        string res;
        for (; n; n /= 16) {
            int tmp = n % 16;
            if (tmp < 10) {
                res += (tmp + '0');
            } else {
                res += (tmp - 10 + 'a');
            }
        }
        return res;
    }

    int hex2Decimal(string s) {
        int res = 0;
        for (char ch : s) {
            int n = (ch >= 'a' ? ch - 'a' + 10 : ch - '0');
            res = res * 16 + n;
        }
        return res;
    }
};